### PR TITLE
[Upgrades testing] Add cases for unchanged key type with changed key expression

### PR DESCRIPTION
**Pull request by moisesackerman-da**
https://docs.google.com/spreadsheets/d/1_0ER--x-YJzoYSoy7VFs6emzn4ZYjBc0kte4MLosuxs/edit?gid=1834568433#gid=1834568433&range=C11

### CODE REVIEW

File: sdk/daml-script/test/daml/upgrades/ContractKeys.daml [hunk 1]
```diff
@@ -98,6 +102,66 @@ exerciseUpdateKeyUnchanged = test $ do
...
+  keyRes <- queryContractKey a $ V2.ChangedKeyExprKey a False
+  case keyRes of
+    Some (foundCid, foundContract) | show foundCid == show cid && foundContract == V2.ChangedKeyExpr a False -> pure ()
+    _ -> assertFail $ "Didn't find correct contract, expected " <> show (cid, V2.ChangedKeyExpr a False) <> ", got " <> show keyRes
+
+queryKeyChangedExprChangedValue : Test
+queryKeyChangedExprChangedValue = test $ do
+  a <- allocateParty "alice"
+  cid <- a `submit` createExactCmd (V1.ChangedKeyExpr a True)
+  -- the following query works, even though the key value changed!
```
Comment (moisesackerman-da) [inline sdk/daml-script/test/daml/upgrades/ContractKeys.daml:159]: @samuel-williams-da @paulbrauner-da I was surprised to find this, what do you think?

File: sdk/daml-script/test/daml/upgrades/ContractKeys.daml [hunk 2]
```diff
@@ -98,6 +104,83 @@ exerciseUpdateKeyUnchanged = test $ do
...
+  cid <- a `submit` createExactCmd (V1.ChangedKeyExpr a False)
+  (foundCid, foundContract) <- a `submit` createAndExerciseCmd (V2.ChangedKeyExprHelper a) (V2.ChangedKeyExprFetch $ V2.ChangedKeyExprKey a False)
+  foundContract === V2.ChangedKeyExpr a False
+  show foundCid === show cid
+
+fetchKeyChangedExprChangedValue : Test
+fetchKeyChangedExprChangedValue = test $ do
+  a <- allocateParty "alice"
+  cid <- a `submit` createExactCmd (V1.ChangedKeyExpr a True)
+  -- the following query works, even though the key value changed!
```
Comment (moisesackerman-da) [inline sdk/daml-script/test/daml/upgrades/ContractKeys.daml:179]: :S

File: sdk/daml-script/test/daml/upgrades/ContractKeys.daml [hunk 3]
```diff
@@ -98,6 +106,111 @@ exerciseUpdateKeyUnchanged = test $ do
...
+  cid <- a `submit` createExactCmd (V1.ChangedKeyExpr a False)
+  foundContract <- a `submit` createAndExerciseCmd (V2.ChangedKeyExprHelper a) (V2.ChangedKeyExprFetch $ coerceContractId cid)
+  foundContract === V2.ChangedKeyExpr a False
+
+fetchKeyChangedExprChangedValue : Test
+fetchKeyChangedExprChangedValue = test $ do
+  a <- allocateParty "alice"
+  cid <- a `submit` createExactCmd (V1.ChangedKeyExpr a True)
+  {-
+  The fetch inside the following command correctly fails with:
```
Comment (moisesackerman-da) [inline sdk/daml-script/test/daml/upgrades/ContractKeys.daml:187]: @samuel-williams-da @paulbrauner-da, it seems that plain `fetch` _does_ perform the upgrade validation

File: sdk/daml-script/test/daml/upgrades/ContractKeys.daml (hunk 1 above)
Comment (samuel-williams-da) [reply sdk/daml-script/test/daml/upgrades/ContractKeys.daml:169]: This test fails then? as expected?

File: sdk/daml-script/test/daml/upgrades/ContractKeys.daml (hunk 3 above)
Comment (samuel-williams-da) [reply sdk/daml-script/test/daml/upgrades/ContractKeys.daml:187]: I would assert that specific failure in the test. for the cases that are broken, write the correct logic and mark them in the test list

File: sdk/daml-script/test/daml/upgrades/ContractKeys.daml [hunk 4]
```diff
@@ -16,14 +21,44 @@ versions: 2
...
 main = tests
-  [ ("Query an unchanged old key for a new contract", queryKeyUnchanged)
-  , ("ExerciseByKey command an unchanged old key for a new contract", exerciseCmdKeyUnchanged)
-  , ("Fetching an unchanged old key for a new contract", fetchKeyUnchanged)
-  , ("ExerciseByKey in Update an unchanged old key for a new contract", exerciseUpdateKeyUnchanged)
-  , ("Query an Upgraded old key for a new contract", queryKeyUpgraded)
-  , ("ExerciseByKey command an Upgraded old key for a new contract", exerciseCmdKeyUpgraded)
-  , ("Fetching an Upgraded old key for a new contract", fetchKeyUpgraded)
-  , ("ExerciseByKey in Update an Upgraded old key for a new contract", exerciseUpdateKeyUpgraded)
+  [ (,) "Unchanged key type" $ tests
```
Comment (moisesackerman-da) [inline sdk/daml-script/test/daml/upgrades/ContractKeys.daml:24]: These were getting quite repetitive so I decided to make the `TestTree` actually a tree, let me know what you think @samuel-williams-da @paulbrauner-da

Comment (samuel-williams-da) [reply sdk/daml-script/test/daml/upgrades/ContractKeys.daml:24]: Either way, looks nice, could consider an alias for `(,)` but thats nitpicking

Comment (moisesackerman-da) [reply sdk/daml-script/test/daml/upgrades/ContractKeys.daml:24]: @samuel-williams-da it works in the sense that the nested test cases run and failure/success is propagated correctly, but the logging isn't great. I've opened https://github.com/digital-asset/daml/pull/19936 to improve that while adding the alias for `subtree`s

**Review by samuel-williams-da — APPROVED** [review]
Very nice! A big chunk of tests for our collection :D

**Pull request merged**